### PR TITLE
Checked locking

### PR DESCRIPTION
**Pull request by mehrdadn**
Ready for merge. Note that these checks don't work on Mac because for some reason the thread ID I get is zero for all threads. Hence, they are disabled for thread IDs that are zero.

### CODE REVIEW

**Comment by robertnishihara** [comment]
Some questions:

- This is a lot of code to add to `scheduler.cc`, is that the right place for it? Why not `utils.cc`? (Is the answer that this functionality is specific to the scheduler (e.g., it needs access to the scheduler's locks))
- Why do we need to create new classes instead of adding them to the original `Synchronized` classes?
- It's a bit ugly to have to add the `__LINE__` macro everywhere. One way around this is to make `get` a macro, although I know you don't like that idea :)
- Can you document what the `MySynchronizedPtr` class does and what precisely it checks?

**Comment by mehrdadn** [comment]
1. The only added code is in `SchedulerService::MySynchronizedPtr`, and yes, it's because it's specific to the `SchedulerService` (hence the "my"), since that's the only way to check if locks are in the correct order.

2. Because somewhere we need to keep track of the list of obtained locks for the object under question (in this case, the scheduler), and that scheduler-specific logic can't be in a generic class like `Synchronized`.

3. It just checks that the locks are obtained in the order of the field declarations in the scheduler... but yeah, I can add a comment.

File: src/scheduler.cc [hunk 1]
```diff
@@ -3,16 +3,138 @@
...
+class SchedulerService::MySynchronizedPtr<void> {
+  SchedulerService* me_;
+  size_t order_;
+  unsigned int line_number_;
+  static unsigned long long get_thread_id() {
+    unsigned long long id = 0;
+    std::stringstream ss;
+    ss << std::this_thread::get_id();
+    ss >> id;
+    // This seems to fail on a Mac. I unfortunately can't find a workaround.
```
Comment (robertnishihara) [inline src/scheduler.cc]: Can you elaborate on this comment? What fails? And what is the failure (e.g., does it crash? or does the check not actually check anything?)

Comment (mehrdadn) [inline src/scheduler.cc]: I have -- see the first post in this PR.

Comment (robertnishihara) [inline src/scheduler.cc]: I read it, but I meant in the code.

Comment (mehrdadn) [inline src/scheduler.cc]: Oh

Comment (mehrdadn) [inline src/scheduler.cc]: See update.

**Comment by mehrdadn** [comment]
FYI, I decided to use a `GET` macro so that I could deduplicate code in addition to avoiding the need to put `__LINE__` everywhere. So now it's `auto field_name = GET(field_name_)`.

File: src/scheduler.cc [hunk 2]
```diff
@@ -748,21 +850,21 @@ bool SchedulerService::attempt_notify_alias(ObjStoreId objstoreid, ObjRef alias_
...
   return true;
 }
 
-void SchedulerService::deallocate_object(ObjRef canonical_objref, const SynchronizedPtr<std::vector<RefCount> > &reference_counts, const SynchronizedPtr<std::vector<std::vector<ObjRef> > > &contained_objrefs) {
+void SchedulerService::deallocate_object(ObjRef canonical_objref, const MySynchronizedPtr<std::vector<RefCount> > &reference_counts, const MySynchronizedPtr<std::vector<std::vector<ObjRef> > > &contained_objrefs) {
   // deallocate_object should only be called from decrement_ref_count (note that
   // deallocate_object also recursively calls decrement_ref_count). Both of
   // these methods take reference_counts and contained_objrefs as argumens,
-  // which are obtained by reference_counts_.get() and contained_objrefs_.get(),
+  // which are obtained by GET(reference_counts_, __LINE__) and get(contained_objrefs_),
```
Comment (robertnishihara) [inline src/scheduler.cc]: Should be `GET(reference_counts_) and GET(contained_objrefs_)`

Comment (mehrdadn) [inline src/scheduler.cc]: Whoops! How'd that fall through. Fixed :)

**Comment by robertnishihara** [comment]
Tried it out. Some notes:

- Reversing the order of two locks in `scheduler.cc` causes the following failure on Ubuntu:

    ```
    fatal error occured: Check failed at line 74 in /home/ubuntu/ray/src/scheduler.cc: (order) >= (found->second.first) with message Thread 140008342152960 attempted to lock workers_ on line 423 after exported_reusable_variables_ 
   ```
    On a Mac OS X it caused no failure.

- Out of the box this is a bit slower, but if we add `#define NDEBUG` in `scheduler.h`, then the performance is unchanged. Numbers below from a m3.large node on EC2 for the command

    ```
    %timeit ray.get(example_functions.increment(0))
    ```

    - On master branch: 100 loops, best of 3: 3.03 ms per loop
    - On lock branch: 100 loops, best of 3: 3.43 ms per loop
    - On lock branch with `#define NDEBUG`: 100 loops, best of 3: 3.03 ms per loop

**Pull request merged**